gce_workload_cert_refresh/main.go

```diff
@@ -79,125 +89,62 @@ func getMetadata(ctx context.Context, key string) ([]byte, error) {
-		err := json.Unmarshal(value, &wc)
-		if err != nil {
-			return err
+MANAGED_WORKLOAD_IDENTITY_SPIFFE is of the format:
+spiffe://POOL_ID.global.PROJECT_NUMBER.workload.id.goog/ns/NAMESPACE_ID/sa/MANAGED_IDENTITY_ID
+
+{
+	"status": "OK", // Status of the response,
+	"workloadCredentials": { // Credentials for the VM's trust domains
+		"MANAGED_WORKLOAD_IDENTITY_SPIFFE": {
+			"certificatePem": "-----BEGIN CERTIFICATE-----datahere-----END CERTIFICATE-----",
+			"privateKeyPem": "-----BEGIN PRIVATE KEY-----datahere-----END PRIVATE KEY-----"
 		}
-		wi.WorkloadCredentials[domain] = wc
 	}
-
-	return nil
 }
+*/
 
 // WorkloadCredential represents Workload Credentials in metadata.
 type WorkloadCredential struct {
-	Metadata       Metadata
-	CertificatePem string
-	PrivateKeyPem  string
+	CertificatePem string `json:"certificatePem"`
+	PrivateKeyPem  string `json:"privateKeyPem"`
 }
 
-/*
-metadata key instance/gce-workload-certificates/root-certs
-
-	{
-	 "status": "OK",
-	 "rootCertificates": {
-	  "PROJECT.svc.id.goog": {
-	   "metadata": {
-	    "workload_creds_dir_path": "/var/run/secrets/workload-spiffe-credentials"
-	   },
-	   "rootCertificatesPem": "-----BEGIN CERTIFICATE-----datahere-----END CERTIFICATE-----"
-	  }
-	 }
-	}
-*/
-
-// WorkloadTrustedRootCerts represents Workload Trusted Root Certs in metadata.
-type WorkloadTrustedRootCerts struct {
-	Status           string
-	RootCertificates map[string]RootCertificate
+// WorkloadIdentities represents Workload Identities in metadata.
+type WorkloadIdentities struct {
+	Status              string                        `json:"status"`
+	WorkloadCredentials map[string]WorkloadCredential `json:"workloadCredentials"`
 }
 
-// UnmarshalJSON is a custom JSON unmarshaller for WorkloadTrustedRootCerts
-func (wtrc *WorkloadTrustedRootCerts) UnmarshalJSON(b []byte) error {
-	tmp := map[string]json.RawMessage{}
-	err := json.Unmarshal(b, &tmp)
-	if err != nil {
-		return err
-	}
-
-	if err := json.Unmarshal(tmp["status"], &wtrc.Status); err != nil {
-		return err
-	}
-
-	wtrc.RootCertificates = map[string]RootCertificate{}
-	rcs := map[string]json.RawMessage{}
-	if err := json.Unmarshal(tmp["rootCertificates"], &rcs); err != nil {
-		return err
-	}
-
-	for domain, value := range rcs {
-		rc := RootCertificate{}
-		err := json.Unmarshal(value, &rc)
-		if err != nil {
-			return err
-		}
-		wtrc.RootCertificates[domain] = rc
-	}
+/*
+metadata key instance/gce-workload-certificates/trust-anchors
+
+{
+    "status":  "<status string>" // Status of the response,
+    "trustAnchors": {  // Trust bundle for the VM's trust domains
+        "PEER_SPIFFE_TRUST_DOMAIN_1": {
+            "trustAnchorsPem" : "<Trust bundle containing the X.509 roots certificates>",
+		},
+        "PEER_SPIFFE_TRUST_DOMAIN_2": {
+            "trustAnchorsPem" : "<Trust bundle containing the X.509 roots certificates>",
+        }
+    }
+}
+*/
 
-	return nil
+// TrustAnchor represents a Root Certificate in metadata
```

Comment:
The TrustAnchorsPem can contain non-root, CA certificates as well. In this case, the certificate chain validation will stop at that certificate. Also, the TrustAnchorsPem can contain multiple certificates one after the another in no particular order. Recommend updating the comment to remove Root and clarifying that TrustAnchor can contain multiple certificates.

Reply:
Done